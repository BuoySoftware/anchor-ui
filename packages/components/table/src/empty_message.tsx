import { useTranslation } from "react-i18next";
import { Body } from "@buoysoftware/anchor-typography";
import { useTable, DEFAULT_TABLE_NAMESPACE } from "./use_table";

export const EmptyMessage: React.FC = (): React.ReactElement => {
  const { tNamespace, tableName } = useTable();
  const { t } = useTranslation();

  const translationLookup = [
    `${tNamespace}:${tableName}.empty_message`,
    `${tNamespace}:tables.${tableName}.empty_message`,
    `${DEFAULT_TABLE_NAMESPACE}:${tableName}.empty_message`,
    `${DEFAULT_TABLE_NAMESPACE}:empty_message`,
  ];

  return (
    <Body mt="xl" size="m" textAlign="center" data-testid="empty-table-state">
      {t(translationLookup)}
    </Body>
  );
};
