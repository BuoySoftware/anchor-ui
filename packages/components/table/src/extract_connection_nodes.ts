export interface Edge<NodeData> {
  node: NodeData;
}
export interface Connection<NodeData> {
  edges: Edge<NodeData>[];
}

export function extractConnectionNodes<NodeData>(
  connection?: Connection<NodeData>
): NodeData[] {
  if (!connection) return [];

  return connection.edges.map(({ node }) => node);
}
