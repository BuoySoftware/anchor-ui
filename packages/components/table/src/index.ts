export * from "./connection_table";
export * from "./empty_cell";
export * from "./table";
