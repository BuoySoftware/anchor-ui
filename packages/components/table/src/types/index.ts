import type { PropertyPath } from "lodash";
import type { UniqueIdentifier } from "@dnd-kit/core";
import { BodyProps } from "@buoysoftware/anchor-typography";

// Pagination utility types
export interface Edge<NodeData> {
  node: Record<NodeData>;
}

export interface Connection<NodeData> {
  edges: Edge<NodeData>[];
}

export interface ConnectionWithEdges<EdgeData> {
  edges: EdgeData[];
}

export interface PageInfo {
  startCursor: string | null;
  endCursor: string | null;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  perPage: number;
}

export interface PageableConnection {
  pageInfo: PageInfo;
  totalCount: number;
}

export interface PaginationVariables {
  variables: {
    after?: string;
    before?: string;
    first?: number;
    last?: number;
  };
}

export type FetchMore = (
  fetchMoreOptions: PaginationVariables
) => Promise<unknown>;

export type PaginationData<RecordData> = (
  | Connection<RecordData>
  | ConnectionWithEdges<RecordData>
) &
  PageableConnection;

// Table component Types
export type CellProps = Partial<BodyProps> | undefined;

export type Record<RecordData> = RecordData & { id?: string };

export type RowAction<RecordData> = (rowData: RecordData) => void;

export type TableVariant = "default" | "condensed";

export interface TableCellConfig<RecordData> {
  cellProps?: (rowData: RecordData) => CellProps;
  dataKey: PropertyPath;
  name?: string;
  render?: (rowData: RecordData) => React.ReactNode;
}

export interface TableProps<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  TableRowComponent?: React.FC<TableRow<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  fetchMore?: FetchMore;
  name: string;
  paginationData?: PaginationData<RecordData>;
  recordIdKey: PropertyPath;
  records: Record<RecordData>[];
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
  tNamespace?: string;
}

export interface TableBodyProps<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  TableRowComponent?: React.FC<TableRow<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  recordIdKey: PropertyPath;
  records: Record<RecordData>[];
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
}

export type RenderPlaceholder<RecordData> = (
  rowData: RecordData
) => React.ReactElement | string;

export interface TableRow<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  record: Record<RecordData>;
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
  rowId: string;
}

export interface TableCell<RecordData> {
  cellConfig: TableCellConfig<RecordData>;
  record: RecordData;
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  testId: string;
  variant?: TableVariant;
}

// Sortable table types
export interface SortableRecord extends Object {
  id: UniqueIdentifier;
}

export interface SortableTableCellConfig<RecordData>
  extends TableCellConfig<RecordData> {
  sortHandle?: boolean;
}

export interface SortableTableRow<RecordData>
  extends Omit<TableRow<RecordData>, "cellConfigs"> {
  cellConfigs: SortableTableCellConfig<RecordData>[];
  sortHandle?: boolean;
}

export interface SortableTableCell<RecordData>
  extends Omit<TableCell<RecordData>, "cellConfig"> {
  cellConfig: SortableTableCellConfig<RecordData>;
  sortHandle?: boolean;
}

export interface SortableTableProps<RecordData>
  extends Omit<
    TableProps<RecordData>,
    "cellConfigs" | "TableCellComponent" | "TableRowComponent"
  > {
  cellConfigs: SortableTableCellConfig<RecordData>[];
  TableCellComponent?: React.FC<SortableTableCell<RecordData>>;
  TableRowComponent?: React.FC<SortableTableRow<RecordData>>;
  onSort?: (records: Record<RecordData>[]) => void;
}
