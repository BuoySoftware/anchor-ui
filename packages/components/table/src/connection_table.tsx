import { Table, TableProps } from "./table";
import { Connection, PageableConnection } from "./types";
import { extractConnectionNodes } from "./extract_connection_nodes";

interface OwnProps<RecordData> {
  connection?: Connection<RecordData> & PageableConnection;
}

type ConnectionTableProps<RecordData> = OwnProps<RecordData> &
  Omit<TableProps<RecordData>, "records" | "paginationData">;

export const ConnectionTable = <RecordData,>({
  connection,
  ...tableBuilderProps
}: ConnectionTableProps<RecordData>): React.ReactElement => {
  const records = extractConnectionNodes(connection);

  return (
    <Table
      paginationData={connection}
      records={records}
      {...tableBuilderProps}
    />
  );
};
