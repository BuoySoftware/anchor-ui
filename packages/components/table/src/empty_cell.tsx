import { TableCellConfig } from "./types";

export const EMPTY_CELL: TableCellConfig<any> = {
  dataKey: "",
  render: () => <></>,
};
