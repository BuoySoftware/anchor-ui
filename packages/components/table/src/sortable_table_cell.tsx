import { useSortable } from "@dnd-kit/sortable";
import { useTheme } from "@buoysoftware/anchor-theme"

import sortHandleGraphic from "./sort_handle.svg";
import { SortableRecord, SortableTableCell as GeneralSortableTableCell } from "./types";
import { Td } from "./td";

export type SortableTableCellProps<RecordData> = GeneralSortableTableCell<RecordData>;

export const SortableTableCell = function <RecordData extends SortableRecord>({
  record,
  cellConfig: sortableCellConfig,
  ...tableCellProps
}: SortableTableCellProps<RecordData>): React.ReactElement {
  const { attributes, listeners, setActivatorNodeRef } = useSortable({
    id: record.id
  });
  const { sizes, space } = useTheme();
  const { sortHandle, ...cellConfig } = sortableCellConfig;

  const sortHandleProps = {
    background: `${space.xs} center url(${sortHandleGraphic}) no-repeat`,
    minHeight: "sortHandle",
    minWidth: "sortHandle",
    pl: `calc(${space.xs} + ${sizes.sortHandle} + ${space.xs})`,
    ref: setActivatorNodeRef,
    style: {
      cursor: "grab"
    },
    ...attributes,
    ...listeners
  }

  return (
    <Td
      cellConfig={cellConfig}
      record={record}
      {...(sortHandle ? sortHandleProps : undefined)}
      {...tableCellProps}
    />
  );
};
