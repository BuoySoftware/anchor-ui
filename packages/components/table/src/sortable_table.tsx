import { useEffect, useCallback, useState } from "react";
import {
  DndContext,
  DragEndEvent,
  KeyboardSensor,
  PointerSensor,
  UniqueIdentifier,
  closestCenter,
  useSensor,
  useSensors,
} from "@dnd-kit/core";
import {
  restrictToVerticalAxis,
  restrictToWindowEdges,
} from "@dnd-kit/modifiers";
import {
  SortableContext,
  arrayMove,
  sortableKeyboardCoordinates,
  verticalListSortingStrategy,
} from "@dnd-kit/sortable";

import {
  SortableRecord,
  SortableTableProps as GeneralSortableTable,
} from "./types";
import { SortableTableCell } from "./sortable_table_cell";
import { SortableTableRow } from "./sortable_table_row";
import { Table } from "./table";

type SortableTableProps<RecordData> = GeneralSortableTable<RecordData>;

export const SortableTable = <RecordData extends SortableRecord>({
  onSort,
  records,
  ...tableBuilderProps
}: SortableTableProps<RecordData>): React.ReactElement => {
  const [sortedRecords, setSortedRecords] = useState(records);
  const sensors = useSensors(
    useSensor(PointerSensor),
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates })
  );

  const handleDragEnd = useCallback(
    ({ active, over }: DragEndEvent) => {
      if (!active.id || !over?.id) return;
      if (active.id === over.id) return;

      function indexById(records: RecordData[], id: UniqueIdentifier) {
        const record = records.find((record) => id === record.id);
        return record ? records.indexOf(record) : -1;
      }

      setSortedRecords((sortedRecords) => {
        const oldIndex = indexById(sortedRecords, active.id);
        const newIndex = indexById(sortedRecords, over.id);
        const resortedRecords = arrayMove(sortedRecords, oldIndex, newIndex);

        return resortedRecords;
      });
    },
    [setSortedRecords]
  );

  useEffect(() => {
    if (!onSort) return;

    if (sortedRecords !== records) {
      onSort(sortedRecords);
    }
  }, [onSort, sortedRecords, records]);

  return (
    <DndContext
      collisionDetection={closestCenter}
      modifiers={[restrictToVerticalAxis, restrictToWindowEdges]}
      onDragEnd={handleDragEnd}
      sensors={sensors}
    >
      <SortableContext
        items={sortedRecords}
        strategy={verticalListSortingStrategy}
      >
        <Table
          TableCellComponent={SortableTableCell}
          TableRowComponent={SortableTableRow}
          records={sortedRecords}
          variant="sortable"
          {...tableBuilderProps}
        />
      </SortableContext>
    </DndContext>
  );
};
