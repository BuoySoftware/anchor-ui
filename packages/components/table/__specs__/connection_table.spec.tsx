import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import { ConnectionTable } from "../src/connection_table";
import { Connection, PageableConnection, TableCellConfig } from "../src/types";
import { render } from "../test_utils";

interface Donation {
  __typename: string;
  id: string;
  status: string;
}

const cellConfigs: TableCellConfig<Donation>[] = [
  { dataKey: "id" },
  { dataKey: "status" },
];

const connection: Connection<Donation> & PageableConnection = {
  edges: [
    {
      node: {
        __typename: "Donation",
        id: "1",
        status: "Good",
      },
    },
    {
      node: {
        __typename: "Donation",
        id: "2",
        status: "Bad",
      },
    },
    {
      node: {
        __typename: "Donation",
        id: "3",
        status: "Ugly",
      },
    },
    {
      node: {
        __typename: "Donation",
        id: "4",
        status: "Other",
      },
    },
  ],
  pageInfo: {
    perPage: 4,
    hasNextPage: false,
    hasPreviousPage: false,
    endCursor: null,
    startCursor: null,
  },
  totalCount: 4,
};

describe("<ConnectionTable />", () => {
  const tables = {
    total_label: "Total {{total}}",
    actions: {
      next: "Next",
      previous: "Previous",
    },
    donations: {
      th: {
        id: "ID",
        status: "Status",
      },
    },
  };

  i18n.use(initReactI18next).init({
    lng: "en",
    ns: ["tables"],
    resources: {
      en: {
        tables,
      },
    },
  });

  it("renders a table based on a column configuration", () => {
    const component = render(
      <ConnectionTable
        cellConfigs={cellConfigs}
        connection={connection}
        name="donations"
        recordIdKey="id"
      />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });
});
