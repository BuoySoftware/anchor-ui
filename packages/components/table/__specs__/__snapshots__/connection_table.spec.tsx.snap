// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectionTable /> renders a table based on a column configuration 1`] = `
<DocumentFragment>
  <table
    class="sc-hLBbgP dxfGUx"
    data-testid="donations-table"
  >
    <thead
      class="sc-pyfCe ecJMuD"
    >
      <tr>
        <th
          class="sc-bcXHqe sc-eDvSVe sc-gKPRtg bjZCcm cRpfjC"
          color="text.primary"
          font-size="font-size-100"
          font-weight="600"
        >
          ID
        </th>
        <th
          class="sc-bcXHqe sc-eDvSVe sc-gKPRtg bjZCcm cRpfjC"
          color="text.primary"
          font-size="font-size-100"
          font-weight="600"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody
      class="sc-ftTHYK jdcCIu"
    >
      <tr
        data-testid="tr-1"
      >
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-1-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          1
        </td>
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-1-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Good
        </td>
      </tr>
      <tr
        data-testid="tr-2"
      >
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-2-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          2
        </td>
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-2-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Bad
        </td>
      </tr>
      <tr
        data-testid="tr-3"
      >
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-3-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          3
        </td>
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-3-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Ugly
        </td>
      </tr>
      <tr
        data-testid="tr-4"
      >
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-4-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          4
        </td>
        <td
          class="sc-bcXHqe sc-eDvSVe sc-jSUZER rcJCu lliSHS"
          color="text.primary"
          data-testid="tr-4-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Other
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
